File: tests_wraps/src/main/java/com/github/klee0kai/test/car/di/cachecontrol/gc/CarGcComponent.java
package com.github.klee0kai.test.car.di.cachecontrol.gc;

import com.github.klee0kai.stone.annotations.component.*;
import com.github.klee0kai.test.car.di.cachecontrol.gc.scopes.GcBumperRedScope;
import com.github.klee0kai.test.car.di.cachecontrol.gc.scopes.GcBumperScope;
import com.github.klee0kai.test.car.di.cachecontrol.gc.scopes.GcWheelScope;
import com.github.klee0kai.test.car.di.cachecontrol.gc.scopes.GcWindowScope;

@Component(
        identifiers = {String.class, Integer.class}
)
public abstract class CarGcComponent {

    public abstract BumperGcModule bumpersModule();

    public abstract WheelGcModule wheelsModule();

    public abstract WheelMappedGcModule wheelMappedModule();

    public abstract WheelMultiMappedGcModule wheelMultiMappedModule();

    public abstract WindowGcModule windowsModule();

    public abstract WindowMappedGcModule windowsMappedModule();

    public abstract WindowMultiMappedGcModule windowsMultiMappedModule();

    @RunGc
    @GcAllScope
    public abstract void gcAll();

    @RunGc
    @GcWeakScope
    public abstract void gcWeak();

    @RunGc
    @GcSoftScope
    public abstract void gcSoft();

    @RunGc
    @GcStrongScope
    public abstract void gcStrong();

    @RunGc
    @GcWheelScope
    public abstract void gcWheels();

    @RunGc
    @GcBumperScope
    public abstract void gcBumpers();

    @RunGc
    @GcBumperRedScope
    public abstract void gcRedBumpers();

    @RunGc
    @GcBumperRedScope
    @GcBumperScope
    public abstract void gcRedBumpers2();

    @RunGc
    @GcWindowScope
    public abstract void gcWindows();

    @RunGc
    @GcWindowScope
    @GcWeakScope
    public abstract void gcWeakWindows();

    @RunGc
    @GcWindowScope
    @GcSoftScope
    public abstract void gcSoftWindows();

    @RunGc
    @GcWindowScope
    @GcBumperScope
    public abstract void gcNothing();

    public void gcWindowsAndWheels() {
        gcWindows();
        gcWheels();
    }


}

File: tests_wraps_kotlin/src/main/kotlin/com/github/klee0kai/test/car/di/cachecontrol/gc/CarGcComponent.kt
package com.github.klee0kai.test.car.di.cachecontrol.gc

import com.github.klee0kai.stone.annotations.component.*
import com.github.klee0kai.test.car.di.cachecontrol.gc.scopes.GcBumperRedScope
import com.github.klee0kai.test.car.di.cachecontrol.gc.scopes.GcBumperScope
import com.github.klee0kai.test.car.di.cachecontrol.gc.scopes.GcWheelScope
import com.github.klee0kai.test.car.di.cachecontrol.gc.scopes.GcWindowScope

@Component(identifiers = [String::class, Integer::class])
abstract class CarGcComponent {
    abstract fun bumpersModule(): BumperGcModule?
    abstract fun wheelsModule(): WheelGcModule
    abstract fun wheelMappedModule(): WheelMappedGcModule?
    abstract fun wheelMultiMappedModule(): WheelMultiMappedGcModule
    abstract fun windowsModule(): WindowGcModule?
    abstract fun windowsMappedModule(): WindowMappedGcModule?
    abstract fun windowsMultiMappedModule(): WindowMultiMappedGcModule?

    @RunGc
    @GcAllScope
    abstract fun gcAll()

    @RunGc
    @GcWeakScope
    abstract fun gcWeak()

    @RunGc
    @GcSoftScope
    abstract fun gcSoft()

    @RunGc
    @GcStrongScope
    abstract fun gcStrong()

    @RunGc
    @GcWheelScope
    abstract fun gcWheels()

    @RunGc
    @GcBumperScope
    abstract fun gcBumpers()

    @RunGc
    @GcBumperRedScope
    abstract fun gcRedBumpers()

    @RunGc
    @GcBumperRedScope
    @GcBumperScope
    abstract fun gcRedBumpers2()

    @RunGc
    @GcWindowScope
    abstract fun gcWindows()

    @RunGc
    @GcWindowScope
    @GcWeakScope
    abstract fun gcWeakWindows()

    @RunGc
    @GcWindowScope
    @GcSoftScope
    abstract fun gcSoftWindows()

    @RunGc
    @GcWindowScope
    @GcBumperScope
    abstract fun gcNothing()


    fun gcWindowsAndWheels() {
        gcWindows()
        gcWheels()
    }
}
